### PR TITLE
Improve performance for 17 rules

### LINKED ISSUE #3760: Improve S818 performance: don't rely on SemanticModel
From [this investigation](https://community.sonarsource.com/t/re-net-analysis-time-during-msbuild-increased-from-15-min-to-40-min/32956/3)

LiteralSuffixUpperCase takes 77 seconds. It's a code smell rule, with little value. We should not rely on the SemanticModel for it.

> 77.890    8      SonarAnalyzer.Rules.CSharp.LiteralSuffixUpperCase

### LINKED ISSUE #3762: S1075 improve performance
S1075 (UriShouldNotBeHardcoded) has a big perf footprint.

We should re-think how we use the Semantic Model inside the rule, in order to improve the performance.

From this [investigation](https://community.sonarsource.com/t/re-net-analysis-time-during-msbuild-increased-from-15-min-to-40-min/32956/3), the rule takes 24 seconds to run - 2%.

> 24.055    2      SonarAnalyzer.Rules.CSharp.UriShouldNotBeHardcoded

### LINKED ISSUE #3764: S1313 HardcodedIpAddress : improve performance
S1313 HardcodedIpAddress is a security hotspot that currently has a big perf hit.

From this [investigation](https://community.sonarsource.com/t/re-net-analysis-time-during-msbuild-increased-from-15-min-to-40-min/32956/3), the rule takes  2%.

> 23.584     2      SonarAnalyzer.Rules.CSharp.UriShouldNotBeHardcoded

We should save some time by doing syntax checks before URI parsing.

### LINKED ISSUE #3766: S4583 improve performance
S4583 BeginInvokePairedWithEndInvoke is sometimes an expensive rule.

This can be slightly improved by doing some AST checks before the semantic model checks.

### LINKED ISSUE #3767: S3220 improve performance
S3220 InvocationResolvesToOverrideWithParams is sometimes expensive in terms of perf.

From this [investigation](https://community.sonarsource.com/t/re-net-analysis-time-during-msbuild-increased-from-15-min-to-40-min/32956/3), the rule takes  2%.

> 22.418    2      SonarAnalyzer.Rules.CSharp.InvocationResolvesToOverrideWithParams

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
Related to #3768 (another small change will be made tomorrow)
Fixes #3767
Fixes #3766 !!
Fixes #3764
Fixes #3762
Fixes #3760
Partial fix for #3731 (another small change tomorrow)

### CODE REVIEW

**Comment by andrei-epure-sonarsource** [comment]
@costin-zaharia-sonarsource I still have to fix the code smells, bringing back to in progress

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs [hunk 1]
```diff
@@ -95,20 +95,16 @@ private static void CheckCountCall(SyntaxNodeAnalysisContext context)
             const string CountName = "Count";
 
             var invocation = (InvocationExpressionSyntax)context.Node;
-            if (!(context.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol) ||
-                methodSymbol.Name != CountName ||
-                invocation.ArgumentList == null ||
+            if (invocation.ArgumentList == null ||
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs:98]: I would try to simplify this a bit. Maybe something similar with:
```
var invocation = (InvocationExpressionSyntax)context.Node;
if (invocation.ArgumentList?.Arguments.Count == 0 &&
    invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    context.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol &&
    methodSymbol.Name == CountName &&
    methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T) &&
    context.SemanticModel.GetTypeInfo(memberAccess.Expression).Type.GetMembers(CountName).OfType<IPropertySymbol>().Any())
{
    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, GetReportLocation(invocation),
                                                         string.Format(MessageUseInstead, $"'{CountName}' property")));
}
```

and maybe extract `context.SemanticModel.GetTypeInfo(memberAccess.Expression).Type.GetMembers(CountName).OfType<IPropertySymbol>().Any()` to a method (can be local)

In this way we have a single list with preconditions and avoid all the negated checks. It is also more concise.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EnumerableSumInUnchecked.cs [hunk 2]
```diff
@@ -46,21 +46,20 @@ protected override void Initialize(SonarAnalysisContext context)
...
-                    }
-
                     var expression = invocation.Expression;
                     if (!(expression is MemberAccessExpressionSyntax memberAccess))
                     {
                         return;
                     }
 
-                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
+                    if (IsSumInsideUnchecked(invocation))
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EnumerableSumInUnchecked.cs:55]: Similar with previous cases, the expression body can be changed with:
```
var invocation = (InvocationExpressionSyntax)c.Node;
if (invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    IsSumInsideUnchecked(invocation))
{
    var methodSymbol = c.SemanticModel.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
    if (IsSumOnInteger(methodSymbol))
    {
        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
    }
}
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetTypeWithIsAssignableFrom.cs [hunk 3]
```diff
@@ -92,12 +97,16 @@ protected override void Initialize(SonarAnalysisContext context)
         private static void CheckIfIsExpressionCanBeReplacedByNullCheck(SyntaxNodeAnalysisContext context)
         {
             var binary = (BinaryExpressionSyntax)context.Node;
-            var typeExpression = context.SemanticModel.GetTypeInfo(binary.Left).Type;
-            var typeCastTo = context.SemanticModel.GetTypeInfo(binary.Right).Type;
 
+            var typeExpression = context.SemanticModel.GetTypeInfo(binary.Left).Type;
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetTypeWithIsAssignableFrom.cs:101]: What do you think about using the following snippet instead?
```
if (context.SemanticModel.GetTypeInfo(binary.Left).Type is {} leftType &&
    leftType.IsClass() &&
    context.SemanticModel.GetTypeInfo(binary.Right).Type is {} rightType &&
    rightType.IsClass() &&
    !rightType.Is(KnownType.System_Object) &&
    leftType.DerivesOrImplements(rightType))
{
    context.ReportDiagnosticWhenActive(Diagnostic.Create(rule, binary.GetLocation(), MessageNullCheck));
}
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetTypeWithIsAssignableFrom.cs [hunk 4]
```diff
@@ -129,13 +138,14 @@ private static void CheckIfIsExpressionCanBeReplacedByNullCheck(SyntaxNodeAnalys
...
-            if (!TypeExaminationOnSystemType.IsGetTypeCall(sideA as InvocationExpressionSyntax, context.SemanticModel))
+            var typeSyntax = (sideB as TypeOfExpressionSyntax)?.Type;
+            if (typeSyntax == null)
             {
                 return;
             }
 
-            var typeSyntax = (sideB as TypeOfExpressionSyntax)?.Type;
-            if (typeSyntax == null)
+            if (!sideA.ToString().Contains("GetType") ||
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/GetTypeWithIsAssignableFrom.cs:147]: I think we can move this up since checking for string is cheap and we can do this before retrieving the `sideB` type.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/OrderByRepeated.cs [hunk 5]
```diff
@@ -20,5 +20,27 @@ public void Test()
...
+            new int[] { 1, 2, 3 }.OrderBy(i => i).ThenBy("x");
+            new int[] { 1, 2, 3 }.OrderBy(i => i).OrderBy("x"); // Noncompliant
+            new int[] { 1, 2, 3 }.OrderBy("x").ThenBy(i => i);
+            var array = new int[] { 1, 2, 3 };
+            MyExtensions.OrderBy(MyExtensions.OrderBy(array, "x"), "x");
+            Foo();
+        }
+
+        public void Foo() { }
+
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/OrderByRepeated.cs:38]: pedantic: empty line

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LiteralSuffixUpperCase.cs [hunk 6]
```diff
@@ -28,7 +28,7 @@ public void Test(long ui)
         public void TestOk()
         {
             const uint a = 0U;
-            const long b = 0L;
+            const long b = 100L;
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LiteralSuffixUpperCase.cs:31]: was this change needed?

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InvocationResolvesToOverrideWithParams.cs [hunk 7]
```diff
@@ -62,10 +62,11 @@ protected override void Initialize(SonarAnalysisContext context)
...
         private static void CheckCall(SyntaxNode node, ArgumentListSyntax argumentList, SyntaxNodeAnalysisContext context)
         {
-            if (!(context.SemanticModel.GetSymbolInfo(node).Symbol is IMethodSymbol invokedMethodSymbol) ||
+            if (argumentList == null ||
+                argumentList.Arguments.Count == 0 ||
+                !(context.SemanticModel.GetSymbolInfo(node).Symbol is IMethodSymbol invokedMethodSymbol) ||
                 !invokedMethodSymbol.Parameters.Any() ||
-                !invokedMethodSymbol.Parameters.Last().IsParams ||
-                argumentList == null)
+                !invokedMethodSymbol.Parameters.Last().IsParams)
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InvocationResolvesToOverrideWithParams.cs:69]: We can include in this "if" the condition from next one to remove one extra exit point from this method.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs [hunk 8]
```diff
@@ -48,14 +49,17 @@ protected override void Initialize(SonarAnalysisContext context)
                     var literal = (LiteralExpressionSyntax)c.Node;
                     var text = literal.Token.Text;
 
-                    if (text[text.Length - 1] == 'l' &&
-                        c.SemanticModel.GetTypeInfo(literal).Type.Is(KnownType.System_Int64))
+                    if (text[text.Length - 1] == 'l' && NotUnsigned(text))
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:52]: We can use here [EndsWith(Char)
](https://docs.microsoft.com/en-us/dotnet/api/system.string.endswith?view=net-5.0#System_String_EndsWith_System_Char_)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs [hunk 9]
```diff
@@ -48,14 +49,17 @@ protected override void Initialize(SonarAnalysisContext context)
...
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, Location.Create(literal.SyntaxTree,
                             new TextSpan(literal.Span.End - 1, 1))));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
+
+        // a number may be `2ul` or `2Ul` and it's easier to read
+        private static bool NotUnsigned(string text) =>
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:62]: Nice! Since we subscribe to `NumericLiteralExpression` there is no need for `SemanticModel` calls to get the type :) 

Can we use `text.Length >= 3`. I think that if it's a number it can't start with a char so the condition can't be true for length = 2.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/OrderByRepeated.cs [hunk 10]
```diff
@@ -46,17 +46,22 @@ protected override void Initialize(SonarAnalysisContext context)
                 c =>
                 {
                     var outerInvocation = (InvocationExpressionSyntax)c.Node;
-                    if (!IsMethodOrderByExtension(outerInvocation, c.SemanticModel))
+                    if (!(outerInvocation.Expression is MemberAccessExpressionSyntax memberAccess))
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/OrderByRepeated.cs:49]: We can simplify the lambda body to:
```
var outerInvocation = (InvocationExpressionSyntax)c.Node;
if (outerInvocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    memberAccess.Expression is InvocationExpressionSyntax innerInvocation &&
    IsMethodOrderByExtension(outerInvocation, c.SemanticModel) &&
    (IsMethodOrderByExtension(innerInvocation, c.SemanticModel) || IsMethodThenByExtension(innerInvocation, c.SemanticModel)))
{
    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
}
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs [hunk 11]
```diff
@@ -47,21 +47,22 @@ protected override void Initialize(SonarAnalysisContext context)
                 {
                     var invocation = (InvocationExpressionSyntax)c.Node;
 
-                    if (!(c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol) ||
-                        methodSymbol.Name != "ToCharArray" ||
-                        !methodSymbol.IsInType(KnownType.System_String) ||
-                        methodSymbol.Parameters.Length != 0)
+                    if (!(invocation.Parent is ElementAccessExpressionSyntax) &&
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs:50]: Can be simplified to something similar with:
```
var invocation = (InvocationExpressionSyntax)c.Node;
if ((invocation.Parent is ElementAccessExpressionSyntax || invocation.Parent is ForEachStatementSyntax) &&
    invocation.Expression is MemberAccessExpressionSyntax memberAccess &&
    invocation.ToString().Contains("ToCharArray") &&
    c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol &&
    methodSymbol.Name == "ToCharArray" && methodSymbol.IsInType(KnownType.System_String) &&
    methodSymbol.Parameters.Length == 0)
{
    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, memberAccess.Name.GetLocation()));
}
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs [hunk 12]
```diff
@@ -47,21 +47,22 @@ protected override void Initialize(SonarAnalysisContext context)
...
 
-                    if (!(invocation.Parent is ElementAccessExpressionSyntax) &&
-                        !(invocation.Parent is ForEachStatementSyntax))
+                    if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
                     {
                         return;
                     }
 
-                    if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+                    if (!invocation.ToString().Contains("ToCharArray") ||
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs:61]: This pattern appears in a lot of places, should we try to extract it or a part in order to enable reuse?

Also related to this, we have `KnownMethods` class which has some methods similar with this. Did you consider extracting the name check as a method there?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TypeExaminationOnSystemType.cs [hunk 13]
```diff
@@ -50,13 +50,18 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                        if (!(c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol))
+                        {
+                            return;
+                        }
 
-                    CheckGetTypeCallOnType(invocation, methodSymbol, c);
-                    CheckIsInstanceOfTypeCallWithTypeArgument(invocation, methodSymbol, c);
+                        CheckGetTypeCallOnType(invocation, methodSymbol, c);
+                        CheckIsInstanceOfTypeCallWithTypeArgument(invocation, methodSymbol, c);
+                    }
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TypeExaminationOnSystemType.cs:64]: We can reduce this to:
```
var invocation = (InvocationExpressionSyntax)c.Node;
var invocationString = invocation.ToString();

if ((invocationString.Contains("IsInstanceOfType") || invocationString.Contains("GetType")) &&
    c.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol)
{
    CheckGetTypeCallOnType(invocation, methodSymbol, c);
    CheckIsInstanceOfTypeCallWithTypeArgument(invocation, methodSymbol, c);
}
```

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs [hunk 14]
```diff
@@ -57,27 +57,27 @@ protected HardcodedIpAddressBase(IAnalyzerConfiguration analyzerConfiguration) :
...
-                    literalValue.Split('.').Length != 4)
+                if (HasAttributes(stringLiteral))
                 {
                     return;
                 }
 
-                var variableName = GetAssignedVariableName(stringLiteral);
-                if (variableName != null &&
-                    IgnoredVariableNames.Any(variableName.Contains))
+                if (literalValue == "::" ||
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs:72]: I think this is less CPU intensive than `GetAssignedVariableName` and `HasAttributes` which have to visit the syntax tree to search for identifiers and attributes.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs [hunk 15]
```diff
@@ -103,22 +103,35 @@ protected override void Initialize(SonarAnalysisContext context)
...
                     var addExpression = (TBinaryExpressionSyntax)c.Node;
-                    if (!IsInCheckedContext(addExpression, c.SemanticModel))
-                    {
-                        return;
-                    }
                     var leftNode = GetLeftNode(addExpression);
+
+                    var isInCheckedContext = false;
+                    // SemanticModel calls are expensive
+                    var ranSemanticCheck = false;
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UriShouldNotBeHardcodedBase.cs:110]: We don't need 2 variables for this. This can be nicely done with `Lazy<bool>` and the code will simplify a lot.
```
var isInCheckedContext = new Lazy<bool>(() => IsInCheckedContext(addExpression, c.SemanticModel));
```

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
I think the changes are very nice and I can't wait for the measurement results!

I've suggested though some changes to simplify the code.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LiteralSuffixUpperCase.cs (hunk 6 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/LiteralSuffixUpperCase.cs:31]: it wasn't needed, however the test case has only 1-digit integer numbers, so I wanted to add a multi-digit test case

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs (hunk 8 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:52]: it was added in .NET Standard 2.1, and we're still on 2.0.

Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:52]: Then maybe the [overload with string](https://docs.microsoft.com/en-us/dotnet/api/system.string.endswith?view=netframework-4.8#overloads)? I think it's easier to read.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs (hunk 12 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/RedundantToCharArrayCall.cs:61]: `KnownMethods` operates on symbols, but I extracted it in extension methods.

I think anyway `ToString` has its cost as well, it probably iterates over the AST and calls the sub-nodes `ToString`s ... but at least if we want to change the method and improve it, we'll have a central place to do it

**Comment by andrei-epure-sonarsource** [comment]
@costin-zaharia-sonarsource I addressed the comments.
Please don't merge yet, as based on the perf results we may need to adjust some of the rules changes.

**Comment by costin-zaharia-sonarsource** [comment]
> Please don't merge yet, as based on the perf results we may need to adjust some of the rules changes.

I'll wait then to review the final changes.

**Comment by andrei-epure-sonarsource** [comment]
I looked just at efcore for saving time. I did two runs for 8.14 and two runs for this version.
- AssertionArgsShouldBePassedInCorrectOrder went from 1902/1941 to 1574/1619
- BeginInvokePairedWithEndInvoke went from 33647/34181 to 3140/3159 (nodatime from 1388 to 250)
- CollectionQuerySimplification went from 13019/13032 to 11710/11815
- DebugAssertHasNoSideEffects went from 2145/2175 to 1926/1964 (for n2cms from 364 to 294)
- EnumerableSumInUnchecked went from 2140/2154 to 1554/1516 (for n2cms from 359 to 125)
- !!!!! FieldsShouldBeEncapsulatedInProperties from <0.001 to 107
- GetTypeWithIsAssignableFrom from 3905/3968 to 2750/2591 (for n2cms from 591 to 310)
- InvocationResolvesToOverrideWithParams from 4802 to 4165
- ! LiteralSuffixUpperCase from 237/247 to 236/243 (but it must depend on the number of NumericLiterals in the codebase!)
- OrderByRepeated from 5121 to 1989 (n2cms from 553 to 175, nodatime 613 to 152)
- ! RedundantCast from 1835/1900 to 1871/1870
- RedundantToCharArrayCall from from 3864 to 1611
- ReturnEmptyCollectionInsteadOfNull from 1837 to 901
- TypeExaminationOnSystemType from 3921 to 3186
- UseStringIsNullOrEmpty from 7699 to 5685
- HardcodedIpAddress from 2 to 2
- UriShouldNotBeHardcoded from 350 to 297

**Comment by andrei-epure-sonarsource** [comment]
Total improvement for these rules as tested on the _efcore_ project - from 86,5 seconds to 42,6 seconds

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs [hunk 16]
```diff
@@ -48,14 +49,18 @@ protected override void Initialize(SonarAnalysisContext context)
                     var literal = (LiteralExpressionSyntax)c.Node;
                     var text = literal.Token.Text;
 
-                    if (text[text.Length - 1] == 'l' &&
-                        c.SemanticModel.GetTypeInfo(literal).Type.Is(KnownType.System_Int64))
+                    if (text[text.Length - 1] == 'l' && !ShouldIgnore(text))
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:52]: don't want to use `text.EndsWith("l")` ?

**Review by costin-zaharia-sonarsource — APPROVED** [review]
Nice! 🚀 

Just 2 small comments.

**Comment by andrei-epure-sonarsource** [comment]
I confirm there was no regression in terms of raised issues on the 3 projects I tested on.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs (hunk 16 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/LiteralSuffixUpperCase.cs:52]: as discussed offline, it's readable as is
